spring-batch-infrastructure/src/main/java/org/springframework/batch/item/json/JsonItemReader.java

```diff
@@ -0,0 +1,121 @@
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+import org.springframework.batch.item.ItemStreamReader;
+import org.springframework.batch.item.file.ResourceAwareItemReaderItemStream;
+import org.springframework.batch.item.support.AbstractItemCountingItemStreamItemReader;
+import org.springframework.beans.factory.InitializingBean;
+import org.springframework.core.io.Resource;
+import org.springframework.util.Assert;
+
+/**
+ * {@link ItemStreamReader} implementation that reads Json objects from a
+ * {@link Resource} having the following format:
+ * <p>
+ * <code>
+ * [
+ *    {
+ *       // JSON object
+ *    },
+ *    {
+ *       // JSON object
+ *    }
+ * ]
+ * </code>
+ * <p>
+ *
+ * The implementation is <b>not</b> thread-safe.
+ *
+ * @param <T> the type of json objects to read
+ *
+ * @author Mahmoud Ben Hassine
+ * @since 4.1
+ */
+public class JsonItemReader<T> extends AbstractItemCountingItemStreamItemReader<T> implements
+		ResourceAwareItemReaderItemStream<T>, InitializingBean {
+
+	private static final Log LOGGER = LogFactory.getLog(JsonItemReader.class);
+
+	private Resource resource;
+
+	private JsonObjectReader<T> jsonObjectReader;
+
+	private boolean strict = true;
+
+	/**
+	 * Set the {@link JsonObjectReader} to use to read and map Json fragments to domain objects.
+	 * @param jsonObjectReader the json object reader to use
+	 */
+	public void setJsonObjectReader(JsonObjectReader<T> jsonObjectReader) {
+		this.jsonObjectReader = jsonObjectReader;
+	}
+
+	/**
+	 * In strict mode the reader will throw an exception on
+	 * {@link #open(org.springframework.batch.item.ExecutionContext)} if the
+	 * input resource does not exist.
+	 * @param strict true by default
+	 */
+	public void setStrict(boolean strict) {
+		this.strict = strict;
+	}
+
+	@Override
+	public void setResource(Resource resource) {
+		this.resource = resource;
+	}
+
+	@Override
+	protected T doRead() throws Exception {
+		return jsonObjectReader.read();
+	}
+
+	@Override
+	protected void doOpen() throws Exception {
+		if (!this.resource.exists()) {
+			if (this.strict) {
+				throw new IllegalStateException("Input resource must exist (reader is in 'strict' mode)");
+			}
+			LOGGER.warn("Input resource does not exist " + this.resource.getDescription());
+			return;
+		}
+		if (!this.resource.isReadable()) {
+			if (this.strict) {
+				throw new IllegalStateException("Input resource must be readable (reader is in 'strict' mode)");
+			}
+			LOGGER.warn("Input resource is not readable " + this.resource.getDescription());
+			return;
+		}
+		this.jsonObjectReader.open(this.resource);
+	}
+
+	@Override
+	protected void doClose() throws Exception {
+		this.jsonObjectReader.close();
+	}
+
+	@Override
+	public void afterPropertiesSet() throws Exception {
+		Assert.notNull(this.jsonObjectReader, "The json object reader must not be null.");
+		Assert.notNull(this.resource, "The resource must not be null.");
+	}
```

Comment:
For the `StaxEventItemReader`, both the resource and unmarshaller are mandatory resources but they are not required at construction time (they are checked in `afterPropertiesSet`). I've seen this in many places in the code base like for example the `ValidatingItemProcessor` where I can create an instance with the default constructor while the validator is mandatory. Another example is the `JmsItemReader`: the jmsTemplate is mandatory according to the `InitializingBean` contract but it is not required in the constructor.

Reply:
I think it's more about class design than configuration style: I should not be able to create a class without specifying its required dependencies (fail fast, at construction time) rather than letting me create the class and then throw me an exception when the container calls `afterPropertiesSet`. Here, I used the `InitializingBean` way for consistency with the current code base, but for me, we should always use constructor injection for required dependencies and setter injection for optional ones *regardless of* the configuration style. I couldn't agree more with you, so I fixed that in 493196c39b363448b08aa6ec8a600a8517e403fa.